### ISSUE TITLE
Provide documentation for users to manage multi-tenant databases 

### ISSUE THREAD

**Issue by markdboyd** [cloud-gov/product#1738]
In order to facilitate low-friction development and testing cloud.gov users, we want to test and document an approach for provisioning a multi-tenant database within an organization.

Essentially, we want users to have the ability to provision a database cluster once (which can take a while) and then subsequently only have to create databases within that cluster, which is significantly faster and reduces development time.

The direct impetus for this story is the deprecation of our shared database plans offered by our RDS service broker, which previously offered the ability to manage a multi-tenant database across applications.

Specifically, we are going to explore using [`tinsmith`](https://github.com/blacksmith-community/cf-postgres-tinsmith) as a service-broker than can manage a multi-tenant database for our users.

## Acceptance Criteria
* [ ] GIVEN that deployment and testing of `tinsmith` demonstrates it can be easily used to manage a multi-tenant database \
  AND an application can successfully write to a database created by `tinsmith`
  THEN user-facing documentation will be written to walk through how to use and implement `tinsmith` for organizations

--- 

## Security considerations

- Is there any known risk to recommending `tinsmith` as a third-party service-broker?

## Implementation sketch

- [x] in a test organization/space (personal space in `cloud-gov-operators`), create an RDS cluster through a service broker
- [x] deploy `tinsmith`: https://github.com/blacksmith-community/cf-postgres-tinsmith
- [x] connect `tinsmith` to RDS cluster
- [x] use `tinsmith` service broker to create a new database, demonstrating ability to create databases in multi-tenant 
- [x] write/adapt basic application that writes to a bound database
  - https://github.com/markdboyd/cf-postgres-example-app- 
- [x] configure application to bind to database created by `tinsmith` and deploy
- [x] verify that application works as expected

Assuming that all of the above steps go well and are successful, then documentation should be written for https://github.com/cloud-gov/cg-site as a knowledge base article to give users a walk-through on implementing a multi-tenant database using `tinsmith`

**Comment by markdboyd** [comment]
Here are the steps I've taken so far as part of this work (all of which were done in the `cloud-gov-operators` org and `mark.boyd` space):

1. Create an RDS database: `cf create-service aws-rds small-psql mark-test-db`
1. Deploy `tinsmith` application and bind it to the database:

    ```
    git clone https://github.com/blacksmith-community/cf-postgres-tinsmith
    cd cf-postgres-tinsmith

    # push the code...
    cf push --no-start
    cf bind-service postgres-tinsmith mark-test-db
    ```

1. Update RDS service instance to have "postgresql" tag so that `tinsmith` can find it:

    ```shell
    cf update-service mark-test-db -t "postgresql"
    ```

1. Finish configuring and deploying `tinsmith`

    ```
    # you may want to set some other environment variables at
    # this stage; see "Configuration", below.
    cf set-env postgres-tinsmith SB_BROKER_USERNAME my-broker
    cf set-env postgres-tinsmith SB_BROKER_PASSWORD a-secret

    # start the app
    cf start postgres-tinsmith

    # register this tinsmith as a service broker in CF
    cf create-service-broker postgres-tinsmith my-broker a-secret \
    https://postgres-tinsmith.app.cloud.gov
    cf enable-service-access postgres

    # marvel at your handiwork
    cf marketplace
    ```

1. Create a database in the shared database: `cf create-service postgres shared mark-shared-db`

Right now, the last step was failing, or at least getting permanently stuck at "create in progress".

At @rbogle's suggestion, I checked out the `tinsmith` logs via `cf logs postgres-tinsmith`. Initially, I didn't see much in there to indicate why things were failing, so I started adding some logging. I also noticed that there was seemingly some error handling missing in the `tinsmith` code. Once I added the error handling, I could see an error happening when it tries to initialize its own `broker` database:

```
   2022-04-28T17:18:10.03-0400 [APP/PROC/WEB/0] ERR unable to create broker database: pq: database "ue500vpzpno18cu1" does not exist
```

Curiously, that database name it seemed to be trying to connect is not the database name exposed in the bindings, but the username. So I started to suspect something was amiss with how the code was preparing the PostgreSQL connection information in the code.

I have minimal experience with Go, but after a few cycles of code updates and redeployment, I was able to get the `tinsmith` service broker initializing properly, which I can because the previous error no longer occurs and I manually verified that the `broker` database now exists via `psql`.

Here is the fork of `tinsmith` which has the debugging changes/fixes I've made:

https://github.com/blacksmith-community/cf-postgres-tinsmith/compare/master...markdboyd:fix-broker-init?expand=1

**Comment by markdboyd** [comment]
After further fixes to my fork of the postgres-tinsmith broker, I was able to successfully create a shared database service with it. Then I was also able to succesfully bind the shared database to [a demo app that writes to a database and returns data](https://github.com/markdboyd/cf-postgres-example-app-/tree/refactor-app) to the shared database and confirm that it works correctly.

I also updated my fork of the `postgres-tinsmith` to use `go mod` for dependency management instead of the deprecated `GoDeps`. 

So I think in terms of a proof-of-concept for using `postgres-tinsmith` to manage shared databases, this testing was successful. The next steps are:

- [ ] Make recommended change from @mogul to fall back to `name` or `database` values in the service instance binding environment variables for the database name
- [ ] Update `postgres-tinsmith` fork to have tests of the broker?
- [ ] Submit a PR from my fork of `postgres-tinsmith` back to the upstream repo
- [ ] Write user-facing documentation on how to implement `postgres-tinsmith` themselves to get shared databases
